### DISCUSSION TITLE
Floating point comparisons: inform or protect the user?

### DISCUSSION THREAD

**Discussion by trexfeathers** [SciTools/iris#4744] [General]
I recently promised to look at changing `Constraint`s to use approximate comparisons, to avoid floating point comparison problems (stemming from #4700). I now think there are compelling arguments for not doing this, but to instead lean on / enhance the [existing warning in the user guide](https://scitools-iris.readthedocs.io/en/latest/userguide/subsetting_a_cube.html#cube-extraction).

What does everyone else think? There was a lot of support for changing to approximate equality in a recent team discussion.

---

1. Principle of least surprise - Iris should be consistent with the rest of the scientific Python community.
   * This is a well known programming concept.
   * AFAICT no other scientific Python hides this problem from end users, rather offering specific solutions and/or guidance (`np.isclose()`, `pd.testing.assert_frame_equal(check_exact=False)`).
   * A particular problem since we know many of our users start their Python journey with Iris.
   * This point is similar to our import debate - #4710.
2. The behaviour change won't be limited to `Constraint`s - much of `Constraint` equality uses `Cell` equality.
   * Can we be sure this is desirable?
   * If it is desirable, do we also need to be consistent by implementing approximate equality _throughout_ Iris? This becomes a much bigger question!
3. Further opportunity for confusion/controversy: should we also make similar changes to `>` / `>=` / `<=` / `<` operations on `Cell`s?
   * These are used when users pass functions to `Constraint`s, consistent behaviour might be expected.
   * IMO this would be consistent if we were already modifying `Cell.__eq__()` - see point 2 - but conversation with @bjlittle suggests that some of us might see this as a line to not cross.
4. Code structure increases change of future divergence.
   * `Constraint`s approach equality in several different ways depending on the specific circumstance, so `==` and `in` are used on many different lines, some calling down to `Cell.__eq__()` while others are doing raw comparisons of values.
   * Very easy for future misunderstanding/skim-reading the code and not being consistent with use of exact or approximate equality.
5. Performance reduction
   * The current code makes it clear that performance is a significant concern for `Constraint`s.
   * Approximate equality involves more operations, so will likely be somewhat slower than exact equality.

Further reading:
* https://github.com/SciTools/iris/pull/2062
* [`_CoordConstraint.extract()`](https://github.com/SciTools/iris/blob/ba07f7b3dec18e44bf461a8a73b69016148b4857/lib/iris/_constraints.py#L264-L328)
* [`Cell.__eq__()`](https://github.com/SciTools/iris/blob/ba07f7b3dec18e44bf461a8a73b69016148b4857/lib/iris/coords.py#L1350-L1387)

**Comment by rcomer** [comment]
I know we’ve said we don’t want to proliferate warnings, and in general I am in agreement with that.  But if a cell doesn’t have bounds and `other` is a float, perhaps it _is_ reasonable for cell equality to throw a warning that this comparison is ill-defined?

**Reply by trexfeathers** [reply]
Ugh I didn't even think about the significance of bounds here.

But in some cases isn't one of the bounds identical to the point? What if both those numbers are infinitesimally higher/lower than the `Constraint` value?

**Reply by rcomer** [reply]
😬

**Comment by trexfeathers** [comment]
Team have discussed an alternative API something like `ApproxConstraint`, which explicitly handles the floating point cases.

* Slower, easier version: convert the `Constraint` into a range function `Constraint`
* Faster, trickier version: work out what the _actual_ floating point representation of the provided number should be, accounting for the type of the coordinate in question, then use that value in the `Constraint` instead.

**Reply by rcomer** [reply]
Constraints on `DimCoord`s already find the nearest matching cell if you are matching to anything other than a `Cell`

https://github.com/SciTools/iris/blob/caae45075795c1e13fbb0f04488c3b99f15b60e8/lib/iris/_constraints.py#L308-L321

I've not thought through the details, but it seems like we should be able to use this.

**Reply by trexfeathers** [reply]
When I read this code I assumed this meant that `nearest_neighbour_index()` needs monotonicity to work / work efficiently.

Alternatively, not using this for `AuxCoord`s could just be an oversight.

**Reply by rcomer** [reply]
Yep, there's nothing in the docstring for `nearest_neighbour_index` to suggest it has to be monotonic.  It has to be 1D, but `extract` only works on 1D coordinates anyway...
